File: android/app/src/main/java/com/sovereignnetworkmobile/NativeZhtpApi.kt
package com.sovereignnetworkmobile

import android.util.Log
import com.facebook.react.bridge.*
import kotlinx.coroutines.*
import org.json.JSONObject
import java.net.URL

class NativeZhtpApi(context: ReactApplicationContext) : ReactContextBaseJavaModule(context) {

  companion object {
    private const val TAG = "NativeZhtpApi"
  }

  override fun getName(): String = "NativeZhtpApi"

  @ReactMethod
  fun signIn(
    identityId: String,
    password: String,
    nodeUrl: String,
    promise: Promise
  ) {
    GlobalScope.launch(Dispatchers.Default) {
      try {
        Log.d(TAG, "signIn: identityId=$identityId")

        if (identityId.trim().isEmpty()) {
          promise.reject(
            "INVALID_INPUT",
            "Identity ID cannot be empty"
          )
          return@launch
        }

        if (password.length < 8) {
          promise.reject(
            "INVALID_INPUT",
            "Password must be at least 8 characters"
          )
          return@launch
        }

        val loginUrl = buildQuicUrl(nodeUrl, "/api/v1/identity/login")
        Log.d(TAG, "Sending login request to: $loginUrl")

        val loginPayload = JSONObject().apply {
          put("identity_id", identityId.trim())
          put("password", password)
        }

        val response = makeQuicRequest(loginUrl, "POST", loginPayload.toString(), 30, "authenticated")
        val statusCode = response.first
        val responseBody = response.second

        Log.d(TAG, "signIn response status: $statusCode")

        when (statusCode) {
          200 -> {
            val identity = JSONObject(responseBody)
            val resultMap = Arguments.createMap().apply {
              putString("identityId", identity.optString("identity_id"))
              putString("did", identity.optString("did"))
              putString("displayName", identity.optString("display_name"))
              putString("identityType", identity.optString("identity_type"))
              putString("deviceId", identity.optString("device_id", null))
              putDouble("createdAt", identity.optDouble("created_at", 0.0))

            }
            Log.d(TAG, "✅ signIn successful: ${identity.optString("did")}")
            promise.resolve(resultMap)
          }

          401, 403 -> {
            promise.reject("INVALID_CREDENTIALS", "Invalid credentials")
          }

          404 -> {
            promise.reject("NOT_FOUND", "Identity not found")
          }

          in 500..599 -> {
            promise.reject("SERVER_ERROR", "Server error: $statusCode")
          }

          else -> {
            promise.reject("UNKNOWN", "HTTP $statusCode")
          }
        }
      } catch (e: Exception) {
        Log.e(TAG, "signIn error: ${e.message}", e)
        promise.reject("UNKNOWN", e.message ?: "Unknown error")
      }
    }
  }

  @ReactMethod
  fun createIdentity(
    displayName: String,
    password: String,
    identityType: String,
    nodeUrl: String,
    promise: Promise
  ) {
    Log.d(TAG, "createIdentity delegating to NativeIdentityProvisioning")
    // Delegate to NativeIdentityProvisioning
    promise.reject("NOT_IMPLEMENTED", "Use NativeIdentityProvisioning for createIdentity")
  }

  @ReactMethod
  fun testConnection(
    nodeUrl: String,
    promise: Promise
  ) {
    GlobalScope.launch(Dispatchers.Default) {
      try {
        Log.d(TAG, "testConnection: $nodeUrl")

        val healthUrl = buildQuicUrl(nodeUrl, "/api/v1/protocol/health")
        val response = makeQuicRequest(healthUrl, "GET", null, 5, "public")
        val statusCode = response.first

        Log.d(TAG, "testConnection: Status $statusCode")

        // 200, 401, 403 means node is reachable
        val connected = statusCode in listOf(200, 401, 403)
        promise.resolve(connected)
      } catch (e: Exception) {
        Log.e(TAG, "testConnection error: ${e.message}")
        promise.resolve(false)
      }
    }
  }

  @ReactMethod
  fun getProtocolInfo(
    nodeUrl: String,
    promise: Promise
  ) {
    GlobalScope.launch(Dispatchers.Default) {
      try {
        Log.d(TAG, "getProtocolInfo: $nodeUrl")

        val healthUrl = buildQuicUrl(nodeUrl, "/api/v1/protocol/health")
        val response = makeQuicRequest(healthUrl, "GET", null, 10, "public")
        val statusCode = response.first
        val responseBody = response.second

        if (statusCode == 200) {
          val protocolInfo = JSONObject(responseBody)
          val resultMap = Arguments.createMap().apply {
            putBoolean("success", protocolInfo.optBoolean("success", true))
            putString("protocolVersion", protocolInfo.optString("protocol_version"))

            val network = protocolInfo.optJSONObject("network")
            val networkMap = Arguments.createMap().apply {
              putString("networkId", network?.optString("network_id"))
              putString("consensus", network?.optString("consensus"))
              putDouble("blockHeight", network?.optDouble("block_height", 0.0) ?: 0.0)
              putInt("peerCount", network?.optInt("peer_count", 0) ?: 0)
            }
            putMap("network", networkMap)

            val node = protocolInfo.optJSONObject("node")
            val nodeMap = Arguments.createMap().apply {
              putString("status", node?.optString("status"))
              putDouble("uptime", node?.optDouble("uptime", 0.0) ?: 0.0)
              putInt("latency", node?.optInt("latency", 0) ?: 0)
              putBoolean("synced", node?.optBoolean("synced", false) ?: false)
            }
            putMap("node", nodeMap)
          }
          Log.d(TAG, "✅ getProtocolInfo successful")
          promise.resolve(resultMap)
        } else {
          promise.reject("SERVER_ERROR", "HTTP $statusCode")
        }
      } catch (e: Exception) {
        Log.e(TAG, "getProtocolInfo error: ${e.message}", e)
        promise.reject("UNKNOWN", e.message ?: "Unknown error")
      }
    }
  }

  @ReactMethod
  fun recoverWithSeed(
    seedPhrase: String,
    nodeUrl: String,
    promise: Promise
  ) {
    promise.reject(
      "NOT_FOUND",
      "Recovery endpoint not implemented on node"
    )
  }

  @ReactMethod
  fun recoverWithBackup(
    backupData: String,
    password: String,
    nodeUrl: String,
    promise: Promise
  ) {
    promise.reject(
      "NOT_FOUND",
      "Backup recovery endpoint not implemented on node"
    )
  }

  @ReactMethod
  fun recoverWithSocial(
    guardianIds: ReadableArray,
    nodeUrl: String,
    promise: Promise
  ) {
    promise.reject(
      "NOT_FOUND",
      "Social recovery endpoint not implemented on node"
    )
  }

  private suspend fun makeQuicRequest(
    urlString: String,
    method: String,
    body: String?,
    timeoutSeconds: Int,
    alpn: String
  ): Pair<Int, String> = withContext(Dispatchers.IO) {
    try {
      val headersJson = if (method == "POST") {
        JSONObject().apply { put("content-type", "application/json") }.toString()
      } else {
        "{}"
      }

      val result = NativeQuicBridge.request(
        url = urlString,
        method = method,
        headersJson = headersJson,
        body = body ?: "",
        timeoutSecs = timeoutSeconds,
        insecure = true,
        alpn = alpn
      )

      val statusCode = (result?.get("status") as? Number)?.toInt() ?: 0
      val responseBody = result?.get("body") as? String ?: ""
      Pair(statusCode, responseBody)
    } catch (e: Exception) {
      Log.e(TAG, "QUIC request error: ${e.message}", e)
      throw e
    }
  }

  private fun buildQuicUrl(nodeUrl: String, path: String): String {
    val normalized = nodeUrl.replaceFirst(Regex("^quic://"), "https://")
    val url = URL(normalized)
    val port = if (url.port != -1) url.port else 443
    return "quic://${url.host}:$port$path"
  }
}

File: ios/NativeZhtpApi.swift
import Foundation
import React

// MARK: - Error Codes
enum ZhtpApiError: String {
  case INVALID_CREDENTIALS = "INVALID_CREDENTIALS"
  case NETWORK_TIMEOUT = "NETWORK_TIMEOUT"
  case NOT_FOUND = "NOT_FOUND"
  case INVALID_INPUT = "INVALID_INPUT"
  case UNAUTHORIZED = "UNAUTHORIZED"
  case SERVER_ERROR = "SERVER_ERROR"
  case UNKNOWN = "UNKNOWN"
}

// MARK: - Data Models
struct ZhtpIdentity: Codable {
  let identityId: String
  let did: String
  let displayName: String
  let identityType: String
  let deviceId: String?
  let createdAt: UInt64?

  enum CodingKeys: String, CodingKey {
    case identityId = "identity_id"
    case did
    case displayName = "display_name"
    case identityType = "identity_type"
    case deviceId = "device_id"
    case createdAt = "created_at"
  }
}

struct ZhtpProtocolInfo: Codable {
  let success: Bool?
  let protocolVersion: String?
  let features: [String: Bool]?
  let network: NetworkInfo?
  let node: NodeInfo?
  let error: String?

  enum CodingKeys: String, CodingKey {
    case success
    case protocolVersion = "protocol_version"
    case features
    case network
    case node
    case error
  }
}

struct NetworkInfo: Codable {
  let networkId: String?
  let consensus: String?
  let blockHeight: UInt64?
  let peerCount: Int?

  enum CodingKeys: String, CodingKey {
    case networkId = "network_id"
    case consensus
    case blockHeight = "block_height"
    case peerCount = "peer_count"
  }
}

struct NodeInfo: Codable {
  let status: String?
  let uptime: UInt64?
  let latency: Int?
  let synced: Bool?
}

// MARK: - React Native Module
@objc(NativeZhtpApi)
class NativeZhtpApi: NSObject {
  private let queue = DispatchQueue(label: "com.sovereignnetwork.zhtp-api", qos: .userInitiated)

  @objc
  static func requiresMainQueueSetup() -> Bool {
    return false
  }

  // MARK: - Sign In
  @objc(signIn:password:nodeUrl:resolve:reject:)
  func signIn(
    identityId: String,
    password: String,
    nodeUrl: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    queue.async {
      if #available(iOS 15.0, *) {
        Task {
          await self.performSignInQuicAuthenticated(
            identityId: identityId,
            password: password,
            nodeUrl: nodeUrl,
            resolve: resolve,
            reject: reject
          )
        }
      } else {
        reject("QUIC_UNSUPPORTED", "QUIC requires iOS 15 or later", nil)
      }
    }
  }

  @available(iOS 15.0, *)
  private func performSignInQuicAuthenticated(
    identityId: String,
    password: String,
    nodeUrl: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) async {
    do {
      print("[NativeZhtpApi] signIn: identityId=\(maskIdentifier(identityId))")

      guard !identityId.trimmingCharacters(in: .whitespaces).isEmpty else {
        throw NSError(
          domain: "ZhtpApiError",
          code: -1,
          userInfo: [
            "code": ZhtpApiError.INVALID_INPUT.rawValue,
            "message": "Identity ID cannot be empty"
          ]
        )
      }

      guard password.count >= 8 else {
        throw NSError(
          domain: "ZhtpApiError",
          code: -1,
          userInfo: [
            "code": ZhtpApiError.INVALID_INPUT.rawValue,
            "message": "Password must be at least 8 characters"
          ]
        )
      }

      let trimmedIdentity = identityId.trimmingCharacters(in: .whitespaces)
      var loginPayload: [String: Any] = [
        "password": password
      ]
      if trimmedIdentity.lowercased().hasPrefix("did:") {
        loginPayload["did"] = trimmedIdentity
      } else {
        loginPayload["identity_id"] = trimmedIdentity
      }

      let loginUrl = try quicUrl(nodeUrl: nodeUrl, path: "/api/v1/identity/login")
      let bodyData = try JSONSerialization.data(withJSONObject: loginPayload)
      let bodyString = String(data: bodyData, encoding: .utf8) ?? "{}"

      print("[NativeZhtpApi] Sending QUIC login request (authenticated) to: \(loginUrl)")

      let response = try await quicRequest(
        url: loginUrl,
        method: "POST",
        headers: ["content-type": "application/json"],
        body: bodyString,
        timeout: 30,
        alpn: "authenticated"
      )

      let status = response["status"] as? Int ?? 0
      let responseBody = response["body"] as? String ?? ""
      let responseData = responseBody.data(using: .utf8) ?? Data()

      print("[NativeZhtpApi] signIn response status: \(status)")

      switch status {
      case 200:
        let decoder = JSONDecoder()
        let identity = try decoder.decode(ZhtpIdentity.self, from: responseData)
      print("[NativeZhtpApi] ✅ signIn successful: \(maskIdentifier(identity.did))")

        let resultMap: [String: Any] = [
          "identityId": identity.identityId,
          "did": identity.did,
          "displayName": identity.displayName,
          "identityType": identity.identityType,
          "deviceId": identity.deviceId as Any? ?? NSNull(),
          "createdAt": identity.createdAt as Any? ?? NSNull()
        ]
        resolve(resultMap)

      case 401, 403:
        throw NSError(
          domain: "ZhtpApiError",
          code: -1,
          userInfo: [
            "code": ZhtpApiError.INVALID_CREDENTIALS.rawValue,
            "message": "Invalid credentials"
          ]
        )

      case 404:
        throw NSError(
          domain: "ZhtpApiError",
          code: -1,
          userInfo: [
            "code": ZhtpApiError.NOT_FOUND.rawValue,
            "message": "Identity not found"
          ]
        )

      case 500...599:
        throw NSError(
          domain: "ZhtpApiError",
          code: -1,
          userInfo: [
            "code": ZhtpApiError.SERVER_ERROR.rawValue,
            "message": "Server error: \(status)"
          ]
        )

      default:
        throw NSError(
          domain: "ZhtpApiError",
          code: -1,
          userInfo: [
            "code": ZhtpApiError.UNKNOWN.rawValue,
            "message": "HTTP \(status)"
          ]
        )
      }
    } catch {
      let errorDict = (error as NSError).userInfo
      reject(
        errorDict["code"] as? String ?? ZhtpApiError.UNKNOWN.rawValue,
        errorDict["message"] as? String ?? error.localizedDescription,
        error
      )
    }
  }

  // MARK: - Test Connection
  @objc(testConnection:resolve:reject:)
  func testConnection(
    nodeUrl: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    queue.async {
      if #available(iOS 15.0, *) {
        Task {
          do {
            print("[NativeZhtpApi] testConnection: \(nodeUrl)")
            let healthUrl = try self.quicUrl(nodeUrl: nodeUrl, path: "/api/v1/protocol/health")
            let response = try await NativeQuic().requestBytes(
              url: healthUrl,
              method: "GET",
              headers: [:],
              body: nil,
              timeout: 5,
              insecure: true,
              alpn: .publicContent
            )
            print("[NativeZhtpApi] testConnection: Status \(response.status)")
            let connected = [200, 401, 403].contains(response.status)
            resolve(connected)
          } catch {
            print("[NativeZhtpApi] testConnection: ERROR - \(error.localizedDescription)")
            resolve(false)
          }
        }
      } else {
        reject("QUIC_UNSUPPORTED", "QUIC requires iOS 15 or later", nil)
      }
    }
  }

  // MARK: - Get Protocol Info
  @objc(getProtocolInfo:resolve:reject:)
  func getProtocolInfo(
    nodeUrl: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    queue.async {
      if #available(iOS 15.0, *) {
        Task {
          await self.performGetProtocolInfoQuic(
            nodeUrl: nodeUrl,
            resolve: resolve,
            reject: reject
          )
        }
      } else {
        reject("QUIC_UNSUPPORTED", "QUIC requires iOS 15 or later", nil)
      }
    }
  }

  @available(iOS 15.0, *)
  private func performGetProtocolInfoQuic(
    nodeUrl: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) async {
    print("[NativeZhtpApi] getProtocolInfo: \(nodeUrl)")
    do {
      let healthUrl = try quicUrl(nodeUrl: nodeUrl, path: "/api/v1/protocol/health")
      let response = try await NativeQuic().requestBytes(
        url: healthUrl,
        method: "GET",
        headers: [:],
        body: nil,
        timeout: 10,
        insecure: true,
        alpn: .publicContent
      )

      if response.status == 200 {
        let decoder = JSONDecoder()
        let protocolInfo = try decoder.decode(ZhtpProtocolInfo.self, from: response.body)
        print("[NativeZhtpApi] ✅ getProtocolInfo successful")

        var networkMap: [String: Any] = [
          "networkId": protocolInfo.network?.networkId as Any? ?? NSNull(),
          "consensus": protocolInfo.network?.consensus as Any? ?? NSNull(),
          "blockHeight": protocolInfo.network?.blockHeight as Any? ?? NSNull(),
          "peerCount": protocolInfo.network?.peerCount as Any? ?? NSNull()
        ]

        var nodeMap: [String: Any] = [
          "status": protocolInfo.node?.status as Any? ?? NSNull(),
          "uptime": protocolInfo.node?.uptime as Any? ?? NSNull(),
          "latency": protocolInfo.node?.latency as Any? ?? NSNull(),
          "synced": protocolInfo.node?.synced as Any? ?? NSNull()
        ]

        let resultMap: [String: Any] = [
          "success": protocolInfo.success ?? true,
          "protocolVersion": protocolInfo.protocolVersion as Any? ?? NSNull(),
          "features": protocolInfo.features as Any? ?? NSNull(),
          "network": networkMap,
          "node": nodeMap
        ]
        resolve(resultMap)
      } else {
        throw NSError(
          domain: "ZhtpApiError",
          code: -1,
          userInfo: [
            "code": ZhtpApiError.SERVER_ERROR.rawValue,
            "message": "HTTP \(response.status)"
          ]
        )
      }
    } catch {
      let errorDict = (error as NSError).userInfo
      reject(
        errorDict["code"] as? String ?? ZhtpApiError.UNKNOWN.rawValue,
        errorDict["message"] as? String ?? error.localizedDescription,
        error
      )
    }
  }

  // MARK: - Recover with Seed (stub - endpoint doesn't exist)
  @objc(recoverWithSeed:nodeUrl:resolve:reject:)
  func recoverWithSeed(
    seedPhrase: String,
    nodeUrl: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    queue.async {
      reject(
        ZhtpApiError.NOT_FOUND.rawValue,
        "Recovery endpoint not implemented on node",
        nil
      )
    }
  }

  // MARK: - Recover with Backup (stub - endpoint doesn't exist)
  @objc(recoverWithBackup:password:nodeUrl:resolve:reject:)
  func recoverWithBackup(
    backupData: String,
    password: String,
    nodeUrl: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    queue.async {
      reject(
        ZhtpApiError.NOT_FOUND.rawValue,
        "Backup recovery endpoint not implemented on node",
        nil
      )
    }
  }

  // MARK: - Recover with Social (stub - endpoint doesn't exist)
  @objc(recoverWithSocial:nodeUrl:resolve:reject:)
  func recoverWithSocial(
    guardianIds: [String],
    nodeUrl: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    queue.async {
      reject(
        ZhtpApiError.NOT_FOUND.rawValue,
        "Social recovery endpoint not implemented on node",
        nil
      )
    }
  }

  // MARK: - Create Identity (NOT IMPLEMENTED - use NativeIdentityProvisioning directly)
  @objc(createIdentity:password:identityType:nodeUrl:resolve:reject:)
  func createIdentity(
    displayName: String,
    password: String,
    identityType: String,
    nodeUrl: String,
    resolve: @escaping RCTPromiseResolveBlock,
    reject: @escaping RCTPromiseRejectBlock
  ) {
    queue.async {
      print("[NativeZhtpApi] createIdentity stub - call NativeIdentityProvisioning.provisionIdentity directly from JavaScript")
      reject(
        ZhtpApiError.NOT_FOUND.rawValue,
        "createIdentity not available here - use NativeIdentityProvisioning.provisionIdentity from JavaScript",
        nil
      )
    }
  }

  private func quicUrl(nodeUrl: String, path: String) throws -> String {
    let normalized = nodeUrl.replacingOccurrences(of: "quic://", with: "https://")
    guard let components = URLComponents(string: normalized), let host = components.host else {
      throw NSError(domain: "ZhtpApiError", code: -1, userInfo: [NSLocalizedDescriptionKey: "Invalid node URL"])
    }
    let port = components.port ?? 443
    return "quic://\(host):\(port)\(path)"
  }

  @available(iOS 15.0, *)
  private func quicRequest(
    url: String,
    method: String,
    headers: [String: String],
    body: String?,
    timeout: Double,
    alpn: String
  ) async throws -> [String: Any] {
    try await withCheckedThrowingContinuation { continuation in
      let options: NSDictionary = [
        "method": method,
        "headers": headers,
        "body": body ?? "",
        "timeout": timeout,
        "insecure": true,
        "alpn": alpn
      ]
      NativeQuic().request(url, options: options, resolve: { result in
        if let map = result as? [String: Any] {
          continuation.resume(returning: map)
        } else {
          continuation.resume(returning: [:])
        }
      }, reject: { _, message, error in
        let err = error ?? NSError(domain: "NativeQuic", code: -1, userInfo: [NSLocalizedDescriptionKey: message ?? "QUIC request failed"])
        continuation.resume(throwing: err)
      })
    }
  }

  private func maskIdentifier(_ value: String?) -> String {
    let trimmed = (value ?? "").trimmingCharacters(in: .whitespaces)
    if trimmed.isEmpty { return "<empty>" }
    let core = trimmed.replacingOccurrences(of: "^did:[^:]*:", with: "", options: .regularExpression)
    if core.count <= 8 { return core }
    let start = core.prefix(4)
    let end = core.suffix(4)
    return "\(start)…\(end)"
  }
}
